core, avm2: Implement AS3 Socket

sleepycatcoding opened the pull request:
Based on #8188

Things left to do:

- [x] Use async connect in desktop
- [ ] Use correct exceptions/errors in Socket methods
- [x] Add command line flag to prevent, ask or always allow Socket connections
- [x] Implement all read*() and write*() methods
- [ ] Web implementation

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,306 @@
+        } else if &endian == b"littleEndian" {
+            socket.set_endian(Endian::Little);
+        } else {
+            return Err(make_error_2008(activation, "value"));
```
nit: this should be `make_error_2008(activation, "endian")`

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,306 @@
+        None => return Ok(Value::Undefined),
+    };
+
+    let host = match args.get(0) {
```
This should be `let host = args.get_value(0);`, since the first parameter is guaranteed.

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+        // We throw an IOError when socket is not open.
+        let handle = socket
+            .get_handle()
+            .ok_or(Error::AvmError(io_error(activation, "TODO", 0)?))?;
```
`io_error(activation, "Error #2002: Operation attempted on invalid socket.", 2002)`

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+        let UpdateContext { sockets, .. } = &mut activation.context;
+
+        // FIXME: Throw correct IoError
+        let handle = socket.get_handle().unwrap();
```
`io_error(activation, "Error #2002: Operation attempted on invalid socket.", 2002)`
Seeing as this error is going to appear often, it might be a good idea to create a function like:
```
fn invalid_socket_error<'gc>(
    activation: &mut Activation<'_, 'gc>,
) -> Error<'gc> {
    io_error(activation, "Error #2002: Operation attempted on invalid socket.", 2002)
}
```

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        let bytearray = args
```
`let bytearray = args.get_object(activation, 0)?;`

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+            .get(0)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_object(activation)?;
+        let offset = args
```
`let offset = args.get_u32(activation, 1)?;`

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+            .get(1)
+            .unwrap_or(&Value::Integer(0))
+            .coerce_to_u32(activation)? as usize;
+        let length = args
```
`let offset = args.get_u32(activation, 2)?;`

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,405 @@
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        let byte = args
```
`let byte = args.get_i32(activation, 0)?;`

Lord-McSweeney commented:
The rest of the parameter lookups should also use `ParametersExt`.

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -272,11 +260,7 @@ pub fn write_byte<'gc>(
-            .cloned()
-            .unwrap_or(Value::Undefined)
-            .coerce_to_i32(activation)?;
+        let byte = args.get_i32(activation, 0)?;
```
nit: Use `get_u32` here.

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -14,18 +15,8 @@ pub fn connect<'gc>(
-        .unwrap_or(&Value::Undefined)
-        .coerce_to_u32(activation)? as u16;
+    let host = args.get_string(activation, 0)?;
+    let port = args.get_i32(activation, 1)? as u16;
```
nit: Use `get_u32` here.

Lord-McSweeney commented:
Apart from my few nits, I think the AVM2-interface part of this looks good. Thank you for contributing!

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,365 @@
+
+        let ba_read = bytearray
+            .as_bytearray()
+            .ok_or("ArgumentError: Parameter must be a bytearray")?;
```
This is guaranteed to be a ByteArray by the AS-side bindings and the `get_object` filtering out any `nill`s, so you can just do `.expect("Parameter must be a bytearray!")`.

Lord-McSweeney commented on core/src/avm2/globals/flash/net/socket.rs:
```diff
@@ -0,0 +1,365 @@
+
+    let host = args.get_string(activation, 0)?;
+    let port = args.get_i32(activation, 1)? as u16;
+
```
This should throw a `SecurityError`: `Error #2003: Invalid socket port number specified.`

sleepycatcoding replied:
I used the TryFrom<u32> impl for u16 here. It errors when u32 is bigger than u16::MAX. Then I mapped the error to the correct SecurityError.

Lord-McSweeney commented on core/src/socket.rs:
```diff
@@ -0,0 +1,239 @@
+                    socket.read_buffer().extend(data);
+
+                    let socket_data_evt =
+                        EventObject::bare_default_event(&mut activation.context, "socketData");
```
This should be a `ProgressEvent`, not an `Event`. According to the Adobe docs:

Event Object Type: [flash.events.ProgressEvent](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/events/ProgressEvent.html)
property ProgressEvent.type = [flash.events.ProgressEvent.SOCKET_DATA](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/events/ProgressEvent.html#SOCKET_DATA)

Lord-McSweeney commented on core/src/socket.rs:
```diff
@@ -0,0 +1,239 @@
+#[derive(Collect)]
+#[collect(no_drop)]
+struct Socket<'gc> {
+    target: Object<'gc>,
```
Wouldn't it be better to have this as an `SocketObject`, since there's no way it could be any other kind of object?

sleepycatcoding replied:
Good idea. I did not think of that.

sleepycatcoding commented on core/src/socket.rs:
```diff
@@ -0,0 +1,217 @@
+                                &mut activation,
+                                &[
+                                    "ioError".into(),
+                                    false.into(),
+                                    false.into(),
+                                    "TODO".into(),
+                                    0.into(),
```
Not sure what message or id to use here.

Lord-McSweeney replied:
For now this can use `Error #2031: Socket Error.`

Lord-McSweeney commented on tests/tests/util/navigator.rs:
```diff
@@ -125,4 +127,17 @@ impl NavigatorBackend for TestNavigatorBackend {
+        _receiver: Receiver<Vec<u8>>,
+        sender: Sender<SocketAction>,
+    ) {
+        sender
```
Since this is for the test backend, I think it might be preferable to log a socket connection here, similarly to how it  was done for `navigate_to_url` and `fetch`, so that testing might be a bit easier. For example:
```
if let Some(log) = &self.log {
    log.avm_trace("Navigator::connect_socket");
    log.avm_trace(&format!("    Host: {}; Port: {}", host, port));
}
```

sleepycatcoding commented on desktop/src/backends/navigator.rs:
```diff
@@ -46,6 +60,8 @@ impl ExternalNavigatorBackend {
         open_url_mode: OpenURLMode,
+        socket_allowed: HashSet<String>,
+        socket_behavior: SocketBehavior,
     ) -> Self {
```
This new function triggers clippy::too_many_arguments lint. Should we allow it here?

Lord-McSweeney commented on core/src/avm2/object/socket_object.rs:
```diff
@@ -24,6 +24,8 @@ pub fn socket_allocator<'gc>(
             endian: Cell::new(Endian::Big),
             object_encoding: Cell::new(ObjectEncoding::Amf3),
+            // Default is 20_000 milliseconds or 20 seconds.
+            timeout: Cell::new(20000),
```
For technical correctness with Flash, the timeout is set when instantiating the Socket, not when creating the object.  So the "default" should be `0`, and the constructor should do this:
```
        public function Socket(host:String = null, port:int = 0) {
            this.timeout = 20000; <-- timeout gets set here
            if (host != null) {
                this.connect(host, port);
            }
        }
```

sleepycatcoding replied:
Done.

Lord-McSweeney commented on desktop/src/backends/navigator.rs:
```diff
@@ -277,4 +294,145 @@ impl NavigatorBackend for ExternalNavigatorBackend {
+                    stream
+                }
+                Err(err) => {
+                    warn!("Failed to connect to {}:{}, error: {}", host2, port, err);
```
AFAIK a timed out connection will throw a SecurityErrorEvent in Flash. It's probably not necessary to implement that right now, but it might be good to add a TODO notice here for that.

sleepycatcoding replied:
Is there a specific error message, id?

sleepycatcoding replied:
I added a way to differentiate between a failed and timed out connection.

sleepycatcoding commented on desktop/src/backends/navigator.rs:
```diff
@@ -277,4 +294,145 @@ impl NavigatorBackend for ExternalNavigatorBackend {
+                match stream
+                    .read(&mut buffer)
+                    .or(async {
+                        Timer::after(Duration::from_millis(50)).await;
```
The timeout here is for reading, not connecting. The problem is some protocols might not send us data before we send them data. As such if that timeout was not there, we would get into a infinite loop waiting for data from server.

sleepycatcoding replied:
I reread your comment and I also don't really know if the timeout should be increased.

sleepycatcoding commented:
> Is there a specific reason `timeout` is an `Option<Duration>` on the backend side? It seems unnecessary.

There is no specific reason. I removed Option from Duration.

Lord-McSweeney commented:
Transformice (#8937) fails to connect with this PR. It works in FP, so the server must still be up. It might be worth looking into that.

sleepycatcoding commented:
It is very interesting as there seem to be two versions of the SWF. I could connect and play with this version https://www.transformice.com/TransformiceChargeur.swf, but the link in the linked issue https://www.transformice.com/Transformice.swf fails with message `Invalid version, try to reload the game.`, this happens both on Ruffle and Flash Player Projector, which leads me to believe they have updated the game since. The link they have on their website for the Old Browser based version (https://www.transformice.com/transformice-swf.html) uses the former SWF, not the latter.

sleepycatcoding commented:
Rebased.

Lord-McSweeney commented:
@sleepycatcoding you are correct. https://www.transformice.com/TransformiceChargeur.swf is the loader for the game, and the game has some magic values that are frequently changed in https://www.transformice.com/Transformice.swf; a different file is served every time. Therefore, the second one is not expected to work.

sleepycatcoding commented:
@Lord-McSweeney I think the problem you might have had with this is that by default socket connections are denied. You can enable them with `--socket unrestricted` or `--socket ask` flags or use the advanced open dialog which has socket behavior under network settings.

Lord-McSweeney commented:
I think the `--socket` option should match the `--open-url-mode` option (`--socket-mode allow/confirm/deny`), though I'd like to get others to weigh in on this. Also, I think setting the default to `deny` might be confusing to users- IMO, a default of `ask` would be better.

Dinnerbone reviewed (approved):
This is really cool. The code looks good, but, I'm a little hesitant about the lack of testing. However, I recognise there's no easy way to test this currently.

I've been pondering on that for a bit, and I think if we:
1. Make a new file `socket.toml` or something, which represents a list of what a mocked up socket should do. Receive X, send X, wait X, disconnect, Any deviation is an error.
2. Write a small utility that hosts a socket locally and runs that script, printing any deviation or progress it makes. Use this to get Flash output.
3. Add an option to the test config to mock a connection using this file, so it doesn't actually open sockets and introduce unpredictability to the tests. The mock socket would run through the script and fail the test if it didn't complete to the end, or received anything out of the ordinary.

I'm happy to not block this PR on that because this is really cool and something is better than nothing, but I think I'd want that for future improvements to this area. Does that sound like something you'd be able to do, or will you want help?

Thank you again for this PR!

torokati44 commented:
> I think the --socket option should match the --open-url-mode option (--socket-mode allow/confirm/deny), though I'd like to get others to weigh in on this.

Agree. Naming it simply `socket` implies that the option itself would name a socket of some sort, which is not the case.

> Also, I think setting the default to deny might be confusing to users (because the connection silently fails)- IMO, a default of ask would be better.

Also agree. Keeping it silently on deny by default diminishes its out-of-the-box usability and discoverability of it's presence. I'd even go as far as to suggest setting it to allow by default on desktop.

Dinnerbone commented:
> Also agree. Keeping it silently on deny by default diminishes its out-of-the-box usability and discoverability of it's presence. I'd even go as far as to suggest setting it to allow by default on desktop.

I think one day when we know that we're secure and following whatever security practices existed at the time, but not today. Ask is a good compromise.

sleepycatcoding commented:
> Agree. Naming it simply `socket` implies that the option itself would name a socket of some sort, which is not the case.

What naming should we use for this? Maybe something along the lines like `socket-behavior` or `socket-mode`?

sleepycatcoding commented:
> This is really cool. The code looks good, but, I'm a little hesitant about the lack of testing. However, I recognise there's no easy way to test this currently.
> 
> I've been pondering on that for a bit, and I think if we:
> 
>     1. Make a new file `socket.toml` or something, which represents a list of what a mocked up socket should do. Receive X, send X, wait X, disconnect, Any deviation is an error.
> 
>     2. Write a small utility that hosts a socket locally and runs that script, printing any deviation or progress it makes. Use this to get Flash output.
> 
>     3. Add an option to the test config to mock a connection using this file, so it doesn't actually open sockets and introduce unpredictability to the tests. The mock socket would run through the script and fail the test if it didn't complete to the end, or received anything out of the ordinary.
> 
> 
> I'm happy to not block this PR on that because this is really cool and something is better than nothing, but I think I'd want that for future improvements to this area. Does that sound like something you'd be able to do, or will you want help?

I would prefer to have the testing implementation in a separate PR.

Hadaward commented:
I was testing this PR just now with Transformice and found that for some reason my ping was much higher in Ruffle than in Flash, something is slowing down the socket.

<img src='https://cdn.discordapp.com/attachments/1083823102758240336/1132207642660585492/image.png' alt=''/>
In flash the ping is 305, in ruffle it is 396, I used the /ping command of the game itself to verify. I also noticed a delay in synchronizing player movement. I tested it at least 10 times, and as much as my ping varied, it was never the same or close in the ruffle, always keeping a good distance between the loaders.

sleepycatcoding commented:
@Hadaward Can you retry with the latest commit? On my machine the change improved the latency difference to about 10 ms.

Lord-McSweeney commented:
I believe this still needs error checking in all the `readX`/`writeX` methods to check if the socket is actually connected.

sleepycatcoding commented:
> I believe this still needs error checking in all the `readX`/`writeX` methods to check if the socket is actually connected.

Added that behavior, also added a test to check that the IOErrors are correctly thrown.

Hadaward commented:
I just tested with the most recent commit, unfortunately I think my internet is not good. The ping in the ruffle and in the flash player are varying a lot, sometimes it is better in the ruffle than in the flash player, and sometimes it is better in the flash player than in the ruffle. But everything related to networking seems to be working. But regarding the previous delay, I almost don't feel it anymore, ruffle and flash player are behaving almost the same. If there's a delay now, it's because my ping isn't stable.

Lord-McSweeney reviewed (approved):
Looks good to me! Tests can be added in a follow-up PR.
